src/rgw/rgw_notify.cc

```diff
@@ -0,0 +1,128 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*- 
+// vim: ts=8 sw=2 smarttab
+
+#include "rgw_notify.h"
+#include <memory>
+#include <boost/algorithm/hex.hpp>
+#include "rgw_pubsub.h"
+#include "rgw_pubsub_push.h"
+#include "rgw_perf_counters.h"
+#include "common/dout.h"
+
+#define dout_subsys ceph_subsys_rgw
+
+namespace rgw::notify {
+
+// populate record from request
+void populate_record_from_request(const req_state *s, 
+        const ceph::real_time& mtime, 
+        const std::string& etag, 
+        EventType event_type,
+        rgw_pubsub_s3_record& record) { 
+  record.eventVersion = "2.1";
+  record.eventSource = "aws:s3";
+  record.eventTime = mtime;
+  record.eventName = to_string(event_type);
+  record.userIdentity = s->user->user_id.id;    // user that triggered the change
+  record.sourceIPAddress = "";                  // IP address of client that triggered the change: TODO
+  record.x_amz_request_id = s->req_id;          // request ID of the original change
+  record.x_amz_id_2 = s->host_id;               // RGW on which the change was made
+  record.s3SchemaVersion = "1.0";
+  // configurationId is filled from subscription configuration
+  record.bucket_name = s->bucket_name;
+  record.bucket_ownerIdentity = s->bucket_owner.get_id().id;
+  record.bucket_arn = to_string(rgw::ARN(s->bucket));
+  record.object_key = s->object.name;
+  record.object_size = s->obj_size;
+  record.object_etag = etag;
+  record.object_versionId = s->object.instance;
+  // use timestamp as per key sequence id (hex encoded)
+  const utime_t ts(real_clock::now());
+  boost::algorithm::hex((const char*)&ts, (const char*)&ts + sizeof(utime_t), 
+          std::back_inserter(record.object_sequencer));
+  // event ID is rgw extension (not in the S3 spec), used for acking the event
+  // same format is used in both S3 compliant and Ceph specific events
+  // not used in case of push-only mode
+  record.id = "";
+  record.bucket_id = s->bucket.bucket_id;
+  // pass meta data
+  record.x_meta_map = s->info.x_meta_map;
+}
+
+bool filter(const rgw_pubsub_topic_filter& filter, const req_state* s, EventType event_type) {
+    // if event list exists, and none of the events in the list matches the event type, filter the message
+    if (filter.events.size() && std::find(filter.events.begin(), filter.events.end(), event_type) == filter.events.end()) {
+        return true;
+    }
+    // TODO: add filter by compliant conf: object name, prefix, suffix
+    // TODO: add extra filtering criteria: object size, ToD, metadata, ...
+    return false;
+}
+
+int publish(const req_state* s, 
+        const ceph::real_time& mtime, 
+        const std::string& etag, 
+        EventType event_type,
+        rgw::sal::RGWRadosStore* store) {
+    RGWUserPubSub ps_user(store, s->user->user_id);
```

Comment:
My understanding on this line is that if User X uploaded an object, the notification config will be retrieved from User X even if User X is not the owner of the Bucket.
Is it an issue here or I misunderstood something?